lib/AST/LocalizationFormat.cpp

```diff
@@ -36,8 +37,21 @@ enum LocalDiagID : uint32_t {
 
 struct DiagnosticNode {
   uint32_t id;
-  std::string msg;
+
+  // A diagnostic ID might be available in `.def` and not in YAML, if that's the
+  // case we will be only interested in the ID name to report all of the
+  // `unknownIDs` later. Otherwise we will store the diagnostic message.
+  union {
+    std::string idName;
```

Comment:
I was thinking that `unknownID` would be a union with `id` not a message...

Let's consider a different approach here - instead of trying to map `id` directly while parsing yaml, how about making `id` to be `std::string` as well and adding a separate function like `Optional<LocalDiagID> getDiagnosticID(llvm::StringRef rawID)` which would be used in `readYAML` right before storing data in the `Seq`.

Reply:
I'm sorry I can't quite understand what you mean. The reason I'm mapping `id` directly while parsing because:
a. I will need to use it in ScalarEnumerationTraits. 
b. The `id` will actually tell me whether the DiagID exists in `.def` or not. https://github.com/apple/swift/pull/33337/files#diff-594606b3ff2e7321e3fac5cc36c08dcbR60-R70
That's why I think I have to map `id` directly. I also would love to know how the approach you suggested will be better, thanks! :)

Reply:
The main point I'm trying to make is that we might be doing mapping of ids to early, I think it would be okay to delay it until right before it gets into a `Seq`.

Reply:
Hmm... what about this:
1. Make `DiagnosticNode` as follows:
```cpp
struct DiagnosticNode {
std::string idName;
std::string msg;
};
```
2. Create an array with macro expansion that has the diagnostic IDs as strings.
3. I'll create `Optional<u_int32> getDiagnosticID(llvm::StringRef rawID)` in `LocalizationInput` and in it I'll try to see if there's any match between the `rawID` and `DiagnosticsIDs` array. If there's a match I'll return the index of the `id` in the array.
4. In `readYAML` I'll use the returned value of `getDiagnosticID` to determine whether the `id` exists or not.
5. I'll remove the `ScalarEnumerationTraits`. We won't need it anymore.

Does this make sense to you?

Reply:
Yes, that's exactly what I was talking about. One nit though let's not rename `id` in `DiagnosticNode`.